pkg/dynamiccontroller/dynamic_controller.go

```diff
@@ -306,265 +304,372 @@ func (dc *DynamicController) processNextWorkItem(ctx context.Context) bool {
-	dc.log.Info("Unregistering GVK", "gvr", gvr)
+// reconcileChildrenLocked ensures that reg.childHandlerIDs matches the desired set.
+// Must be called with dc.mu held.
+func (dc *DynamicController) reconcileChildrenLocked(
+	parent schema.GroupVersionResource,
+	desired []schema.GroupVersionResource,
+	reg *registration,
+) error {
+	desiredSet := make(map[schema.GroupVersionResource]struct{}, len(desired))
+	for _, g := range desired {
+		desiredSet[g] = struct{}{}
+	}
+
+	parentGVRKey := keyFromGVR(parent)
+
+	// remove obsolete
+	for child, childHandlerID := range reg.childHandlerIDs {
+		if _, keep := desiredSet[child]; keep {
+			continue
+		}
+		if err := dc.removeHandlerLocked(child, childHandlerID); err != nil {
+			return fmt.Errorf("removing child handler %s: %w", child, err)
+		}
+		delete(reg.childHandlerIDs, child)
+
+		childGVRKey := keyFromGVR(child)
+		handlerDetachTotal.WithLabelValues("child").Inc()
+		handlerCount.WithLabelValues("child").Dec()
+		dc.log.V(1).Info("Detached child", "parent", parentGVRKey, "gvr", childGVRKey)
+	}
 
-	// Retrieve the informer
-	informerObj, ok := dc.informers.Load(gvr)
+	// add missing
+	for child := range desiredSet {
+		if _, exists := reg.childHandlerIDs[child]; exists {
+			continue
+		}
+		w := dc.ensureWatchLocked(child)
+		childHandlerID := childHandlerID(parent, child)
+		if err := w.AddHandler(dc.ctx, childHandlerID, dc.handlerForChildGVR(parent, child)); err != nil {
+			return fmt.Errorf("add child handler %s: %w", child, err)
+		}
+		reg.childHandlerIDs[child] = childHandlerID
+
+		childGVRKey := keyFromGVR(child)
+		handlerAttachTotal.WithLabelValues("child").Inc()
+		handlerCount.WithLabelValues("child").Inc()
+		dc.log.V(1).Info("Attached child", "parent", parentGVRKey, "gvr", childGVRKey)
+	}
+
+	return nil
+}
+
+// removeHandlerLocked removes a handler with the specified ID for a given GroupVersionResource watch.
+// If the handler removal stops the watch, the corresponding GroupVersionResource entry is deleted from the watches map.
+// Returns an error if the handler removal fails.
+func (dc *DynamicController) removeHandlerLocked(gvr schema.GroupVersionResource, handlerID string) error {
+	w, ok := dc.watches[gvr]
 	if !ok {
-		dc.log.V(1).Info("GVK not registered, nothing to deregister", "gvr", gvr)
+		// no watch means there cannot be a handler
 		return nil
 	}
+	stopped, err := w.RemoveHandler(handlerID)
+	if err != nil {
+		return fmt.Errorf("removing handler %s for watch on %s failed: %w", handlerID, gvr, err)
+	}
+	if stopped {
+		delete(dc.watches, gvr)
+	}
+	return nil
+}
 
-	wrapper, ok := informerObj.(*informerWrapper)
-	if !ok {
-		return fmt.Errorf("invalid informer type for GVR: %s", gvr)
+func (dc *DynamicController) handlerForChildGVR(parent, child schema.GroupVersionResource) cache.ResourceEventHandler {
+	parentGVRKey, childGVRKey := keyFromGVR(parent), keyFromGVR(child)
+	handle := func(obj interface{}, eventType string) {
```

Comment:
minor, style, feel free to ignore: Could this be a method on `DynamicController` too rather than inline here?